### PR TITLE
Improve unit test cases for annotation with readonly types

### LINKED ISSUE #24245: Improve unit test cases for annotation with readonly types
**Description:**

- [ ] add readonly type test cases
There are already usages for read-only values 
refer comment: https://github.com/ballerina-platform/ballerina-lang/pull/23473#pullrequestreview-432068262
Object read-only check for annotations is skipped for now. This issue is created to add unit test cases for other scenarios.

- [ ] We can do regex expression check for unit test cases from ballerina side as well.
https://github.com/ballerina-platform/ballerina-lang/pull/23473#discussion_r441287745

- [ ] Fix PR suggestion:
https://github.com/ballerina-platform/ballerina-lang/pull/23473#discussion_r441288061

- [ ] Add a test case to reflect runtime changes.
We have override copy, freezeDirect like methods to allow Abstract objects for cloneReadOnly.

```Ballerina
type Annot record {|
   service s;
   object {} t;
   handle u;
   typedesc v;
   function () w;
|};
```

**Related Issues (optional):**
#24217 (objects are skipped for now)
#24246 (enable readonly object checks for annotation value)

**Suggested Labels (optional):**
Area/Language

**Suggested Assignees (optional):**
KRV/Perera

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
Fixes #24245

- [x] add readonly type test cases
There are already usages for read-only values 
refer comment: https://github.com/ballerina-platform/ballerina-lang/pull/23473#pullrequestreview-432068262
Object read-only check for annotations is skipped for now. This issue is created to add unit test cases for other scenarios.

- [x] We can do regex expression check for unit test cases from ballerina side as well.
https://github.com/ballerina-platform/ballerina-lang/pull/23473#discussion_r441287745

- [x] Fix PR suggestion:
https://github.com/ballerina-platform/ballerina-lang/pull/23473#discussion_r441288061

- [x] Add a test case to reflect runtime changes.
We have override copy, freezeDirect like methods to allow Abstract objects for cloneReadOnly.

## Remarks

There was a crash which also gets fixed by this.

```
[2020-06-17 12:34:27,745] SEVERE {b7a.log.crash} - function () cannot be immutable
org.ballerinalang.jvm.util.exceptions.BallerinaException: function () cannot be immutable
	at org.ballerinalang.jvm.types.BType.getImmutableType(BType.java:157)
	at org.ballerinalang.jvm.values.ReadOnlyUtils.setImmutableIntersectionType(ReadOnlyUtils.java:100)
	at org.ballerinalang.jvm.values.ReadOnlyUtils.getImmutableType(ReadOnlyUtils.java:95)
	at org.ballerinalang.jvm.values.ReadOnlyUtils.setImmutableIntersectionType(ReadOnlyUtils.java:172)
	at org.ballerinalang.jvm.values.ReadOnlyUtils.setImmutableTypeAndGetEffectiveType(ReadOnlyUtils.java:87)
	at org.ballerinalang.jvm.values.MapValueImpl.freezeDirect(MapValueImpl.java:435)
	at org.ballerinalang.jvm.values.MapValueImpl.frozenCopy(MapValueImpl.java:405)
	at org.ballerinalang.jvm.values.CloneUtils.cloneReadOnly(CloneUtils.java:67)
	at org.ballerinalang.langlib.value.CloneReadOnly.cloneReadOnly(CloneReadOnly.java:45)
	at ballerina.lang_value.1_0_0.value.cloneReadOnly(value.bal:39)
	at annotation_readonly_types.$annot_func$0(annotation_readonly_types.bal:53)
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal [hunk 1]
```diff
@@ -35,11 +35,21 @@ public type T1 record {
...
+    var fn = function() {
+        typedesc<any> t = typeof a;
+        Annot? annot = t.@v1;
+        if (annot is Annot) {
+            annot.foo = "EDITED";
+        }
+    };
+
+    error? res = trap fn();
+        assertEquality(true, res is error);
```
Comment (rdhananjaya) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal:47]: extra indentation?

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal:47]: need to fix. Thanks

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal]: Fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal [hunk 2]
```diff
@@ -65,9 +85,44 @@ function testAnnotationOnObjectTypeReadonlyValueEdit() {
...
+    error? res = trap fn();
+    assertTrue(res is error);
+
+    error resError = <error> res;
+    assertEquality("{ballerina/lang.map}InvalidUpdate", resError.message());
+    assertEquality("Invalid update of record field: modification not allowed on readonly value",
+                   resError.detail()["message"].toString());
+}
+
+public function main() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal:105]: Do we need this main function?

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_readonly_types.bal [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+T1 a = { name: "John" };
+
+function testReadonlyTypeAnnotationAttachment()  {
+    typedesc<any> t = typeof a;
+    Annot? annot = t.@v1;
+    assertTrue(annot is Annot);
+}
+
+public function main() {
+    testTypeAnnotAccess1();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_readonly_types.bal:73]: Does this compile? Can't seem to find `testTypeAnnotAccess1`.

We don't need this main function anyway right?

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_readonly_types.bal]: No..I have run locally disabling the test suite.
Fixed: 5604d29b2fc131265541d30ea1335c316fb8fe82

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal (hunk 2 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annotation_access_negative.bal]: removed

**Pull request merged**